File: weather_dashboard.js
// weather_dashboard.js

const apiKey = 'YOUR_API_KEY'; // Replace with your OpenWeatherMap API key

/**
 * Fetch current weather based on city name.
 * @param {string} city - Name of the city.
 */
async function getCurrentWeather(city) {
    const response = await fetch(`https://api.openweathermap.org/data/2.5/weather?q=${city}&appid=${apiKey}`);
    if (!response.ok) throw new Error('Weather data not found');
    return await response.json();
}

/**
 * Fetch 5-day weather forecast based on city name.
 * @param {string} city - Name of the city.
 */
async function get5DayForecast(city) {
    const response = await fetch(`https://api.openweathermap.org/data/2.5/forecast?q=${city}&appid=${apiKey}`);
    if (!response.ok) throw new Error('Forecast data not found');
    return await response.json();
}

/**
 * Fetch weather based on geographical coordinates.
 * @param {number} lat - Latitude.
 * @param {number} lon - Longitude.
 */
async function getWeatherByCoordinates(lat, lon) {
    const response = await fetch(`https://api.openweathermap.org/data/2.5/weather?lat=${lat}&lon=${lon}&appid=${apiKey}`);
    if (!response.ok) throw new Error('Weather data not found');
    return await response.json();
}

/**
 * Fetch air quality data based on city name.
 * @param {string} city - Name of the city.
 */
async function getAirQuality(city) {
    const response = await fetch(`https://api.openweathermap.org/data/2.5/air_pollution?q=${city}&appid=${apiKey}`);
    if (!response.ok) throw new Error('Air quality data not found');
    return await response.json();
}

/**
 * Fetch weather for multiple cities.
 * @param {Array<string>} cities - An array of city names.
 */
async function getMultipleCitiesWeather(cities) {
    return await Promise.all(cities.map(city => getCurrentWeather(city)));
}

// Example usage:
// getCurrentWeather('London').then(data => console.log(data));
// get5DayForecast('London').then(data => console.log(data));
// getWeatherByCoordinates(51.5074, -0.1278).then(data => console.log(data));
// getAirQuality('London').then(data => console.log(data));
// getMultipleCitiesWeather(['London', 'New York']).then(data => console.log(data));

File: weather_dashboard.py
import requests
import json

class WeatherDashboard:
    def __init__(self, api_key):
        self.api_key = api_key
        self.base_url = "http://api.openweathermap.org/data/2.5/"

    def get_current_weather(self, city):
        endpoint = f"{self.base_url}weather?q={city}&appid={self.api_key}&units=metric"
        response = requests.get(endpoint)
        return response.json()

    def get_5_day_forecast(self, city):
        endpoint = f"{self.base_url}forecast?q={city}&appid={self.api_key}&units=metric"
        response = requests.get(endpoint)
        return response.json()

    def get_weather_by_coordinates(self, lat, lon):
        endpoint = f"{self.base_url}weather?lat={lat}&lon={lon}&appid={self.api_key}&units=metric"
        response = requests.get(endpoint)
        return response.json()

    def get_air_quality(self, lat, lon):
        air_quality_base_url = "http://api.openweathermap.org/data/2.5/air_pollution"
        endpoint = f"{air_quality_base_url}?lat={lat}&lon={lon}&appid={self.api_key}"
        response = requests.get(endpoint)
        return response.json()

# Example usage
if __name__ == '__main__':
    API_KEY = 'YOUR_OPENWEATHERMAP_API_KEY'
    dashboard = WeatherDashboard(API_KEY)
    print(dashboard.get_current_weather('London'))
    print(dashboard.get_5_day_forecast('London'))
    print(dashboard.get_weather_by_coordinates(51.5074, -0.1278))
    print(dashboard.get_air_quality(51.5074, -0.1278))
